### ISSUE TITLE
Perturbation attack results differ significantly from paper

### ISSUE THREAD

**Issue by KaiK-BUW** [haichen-ber/IoU-S-Attack#3]
Hi, and thanks for releasing this code!

I’m using the three attack types described in the paper (Attachment, Detachment, and Perturbation) for some experiments. My results for Attachment and Detachment are very close to those reported in the paper, but the results for Perturbation differ substantially from the expected performance.

For context, I have tested the attacks on both PointPillars and CenterPoint, and in both cases the Perturbation results are consistently off, while Attachment and Detachment behave as expected.

To integrate the attack into my own adversarial attack pipeline, I extracted the relevant attack logic into a standalone function. I was careful to preserve the original functionality and only modified the input/pipeline, not the internal computations of the attack itself.

However, despite confirming that the logic matches the original implementation, the Perturbation attack produces noticeably different results compared to those shown in the paper.

I also experimented with different random start magnitudes for the Perturbation attack. In the end I decided to use: `delta = torch.rand_like(points) * 1e-1`, because the perturbation remained hardly visible while getting better results. However, even with these variations, the results remain far from those shown in the paper.

Are there any subtle implementation details or assumptions specific to the Perturbation attack that might not be obvious from the code?

If it helps, I can share the function!

Thanks in advance for any guidance—really appreciate your work!

**Comment by haichen-ber** [comment]
> Hi, and thanks for releasing this code!
> 
> I’m using the three attack types described in the paper (Attachment, Detachment, and Perturbation) for some experiments. My results for Attachment and Detachment are very close to those reported in the paper, but the results for Perturbation differ substantially from the expected performance.
> 
> For context, I have tested the attacks on both PointPillars and CenterPoint, and in both cases the Perturbation results are consistently off, while Attachment and Detachment behave as expected.
> 
> To integrate the attack into my own adversarial attack pipeline, I extracted the relevant attack logic into a standalone function. I was careful to preserve the original functionality and only modified the input/pipeline, not the internal computations of the attack itself.
> 
> However, despite confirming that the logic matches the original implementation, the Perturbation attack produces noticeably different results compared to those shown in the paper.
> 
> I also experimented with different random start magnitudes for the Perturbation attack. In the end I decided to use: `delta = torch.rand_like(points) * 1e-1`, because the perturbation remained hardly visible while getting better results. However, even with these variations, the results remain far from those shown in the paper.
> 
> Are there any subtle implementation details or assumptions specific to the Perturbation attack that might not be obvious from the code?
> 
> If it helps, I can share the function!
> 
> Thanks in advance for any guidance—really appreciate your work!

Thank you for your attention. We don't have any other special settings, but I suggest you check if the loss is decreasing, and whether the gradient of the point cloud is changing.

**Comment by cadip92** [comment]
+1

**Comment by KaiK-BUW** [comment]
Thank you for your fast reply!

I think the issue may indeed be related to the losses. The adversarial loss is not really improving, and the overall loss also stays almost the same throughout the iterations. Here are the printed losses during optimization:
```
iteration 0, adv_loss: 402.4044, dist_loss: 36.9155235291
Loss all:  439.3199462890625
iteration 100, adv_loss: 408.6970, dist_loss: 3.1806230545
Loss all:  411.8775939941406
iteration 200, adv_loss: 418.5858, dist_loss: 5.4863977432
Loss all:  424.0721740722656
iteration 300, adv_loss: 427.4150, dist_loss: 10.7945079803
Loss all:  438.2095031738281
iteration 400, adv_loss: 427.4338, dist_loss: 12.3988151550
Loss all:  439.8326416015625
iteration 499, adv_loss: 419.8034, dist_loss: 10.3602523804
Loss all:  430.1636657714844
```
The adversarial losses fluctuate but don’t show a meaningful decreasing trend, which is likely why the perturbation attack is not matching the expected performance.

Below is the part of my code where the optimizer and the loss are defined (copied almost 1-to-1 from your repository):
Here is the Adam optimizer initialization:
```
adv_point = adv_point.clone().detach().requires_grad_()
opt = optim.Adam([adv_point], lr=self.attack_lr, weight_decay=0.)
```
And here is the loss computation + optimizer step (within the step-loop):
```
if ious_3d_sorted.shape[1] == 0:
    adv_point = adv_point
else:
    if idx.shape[1] == 0:
        o_bestattack[0] = adv_point.detach().clone()
    else:
        loss = 0
        for j in range(idx.shape[-1]):
            pre_score_truth = pre_score[idx[:, j]]   # shape [9]
            ious_3d = ious_3d_sorted[:, j]           # shape [9]
            if self.sub_loss == 'iou':
                loss_ = -(torch.log(1 - ious_3d + 1e-8))
            elif self.sub_loss == 'score':
                loss_ = -(torch.log(1 - pre_score_truth + 1e-8))
            elif self.sub_loss == 'all':
                loss_ = -(torch.log(1 - pre_score_truth + 1e-8) + torch.log(1 - ious_3d + 1e-8))
            loss += loss_

        # compute dist loss
        dist1 = dist_func(adv_point[:, :3][None, :, :], points[:, :3][None, :, :])
        dist2 = dist_func(points[:, :3][None, :, :], adv_point[:, :3][None, :, :])
        dist_loss = dist1 + dist2

        dist_loss = dist_loss + torch.sqrt(torch.sum((adv_point[:, :3] - points[:, :3]) ** 2, dim=[0, 1]) + 1e-4)
        loss_all = loss.sum() + dist_loss
        opt.zero_grad()
        loss_all.backward()
        opt.step()
```
These sections are essentially identical to your original code, so I’m not sure what could be causing the issue here. 

I also checked the gradients of the point clouds, and they do change over iterations, so I don’t think gradient flow is the problem.

If you have any idea what else could cause the adversarial loss to stagnate like this in the Perturbation attack, I’d be grateful for any pointers!

Thank you for your time!

**Comment by haichen-ber** [comment]
> These sections are essentially identical to your original code, so I’m not sure what could be causing the issue here.
> 
> I also checked the gradients of the point clouds, and they do change over iterations, so I don’t think gradient flow is the problem.
> 
> If you have any idea what else could cause the adversarial loss to stagnate like this in the Perturbation attack, I’d be grateful for any pointers!
> 
> Thank you for your time!

The gradient in the point cloud is likely due to distance loss, not adversarial loss. I suggest first removing the distance loss, and then observing the change in loss and whether there is a gradient. In addition, if there is no gradient, it is likely that there are some operations in the model where the gradient is disconnected. As mentioned in our readme, this needs to be carefully debugged and investigated.

**Comment by KaiK-BUW** [comment]
You were right, the issue was indeed related to the gradients and how the losses interacted.

After debugging further, I found that the gradients coming from the adversarial loss after voxelization had a much smaller magnitude compared to the gradients from the distance loss. Because of that imbalance, the optimization was dominated by the distance loss, and the adversarial loss barely had any effect.

To address this, I normalized both gradient terms and applied a scaling factor (×5) to the adversarial-loss gradient so that it contributes more strongly during optimization. I’ve only tested this on a few samples so far, but the results already look much better.

I’m now running the attack across the full nuScenes dataset, though it will take a lot of time. I’ll update you once it finishes.

Thank you for your time and help!

**Comment by haichen-ber** [comment]
> You were right, the issue was indeed related to the gradients and how the losses interacted.
> 
> After debugging further, I found that the gradients coming from the adversarial loss after voxelization had a much smaller magnitude compared to the gradients from the distance loss. Because of that imbalance, the optimization was dominated by the distance loss, and the adversarial loss barely had any effect.
> 
> To address this, I normalized both gradient terms and applied a scaling factor (×5) to the adversarial-loss gradient so that it contributes more strongly during optimization. I’ve only tested this on a few samples so far, but the results already look much better.
> 
> I’m now running the attack across the full nuScenes dataset, though it will take a lot of time. I’ll update you once it finishes.
> 
> Thank you for your time and help!

I suggest you carefully review our README section on handling non-differentiable voxels. Resolving this should prevent future issues. Additionally, please check if the model's precision is float32; this is crucial, otherwise it may lead to gradient explosion or NaN loss.

**Comment by KaiK-BUW** [comment]
Thanks for the suggestion, that resolved the issue.

Following your advice, I reviewed the solution on non-differentiable voxels and updated my implementation to handle them exactly as done in the BEV-Robust repository for the TransFusion code. I also realized that I had indeed forgotten to explicitly set the model precision to float32.

After making these changes and rerunning the experiments, the Perturbation results are now very close to those reported in the paper.

Thank you very much for your time and help! I really appreciate it!
